projects/update_ping_ready_nightly_validation.kp/knowledge.md

```diff
@@ -0,0 +1,347 @@
+    Target buildId mismatch: 20170805100334 must be more recent than 20170809100326:	0.001%
+    Target buildId mismatch: 20170806100257 must be more recent than 20170807113452:	0.002%
+    Target buildId mismatch: 20170806100257 must be more recent than 20170808114032:	0.001%
+    Target buildId mismatch: 20170806100257 must be more recent than 20170810100255:	0.001%
+    Target buildId mismatch: 20170806100257 must be more recent than 20170812100345:	0.001%
+    Target buildId mismatch: 20170807113452 must be more recent than 20170808114032:	0.002%
+    Target buildId mismatch: 20170807113452 must be more recent than 20170809100326:	0.003%
+    Target buildId mismatch: 20170807113452 must be more recent than 20170810100255:	0.002%
+    Target buildId mismatch: 20170808100224 must be more recent than 20170808114032:	0.001%
+    Target buildId mismatch: 20170808114032 must be more recent than 20170809100326:	0.005%
+    Target buildId mismatch: 20170808114032 must be more recent than 20170810100255:	0.001%
+    Target buildId mismatch: 20170809100326 must be more recent than 20170810100255:	0.002%
+    Target buildId mismatch: 20170809100326 must be more recent than 20170811100330:	0.001%
+    Target buildId mismatch: 20170810100255 must be more recent than 20170811100330:	0.001%
+    Target buildId mismatch: 20170810100255 must be more recent than 20170812100345:	0.001%
+    Target channel mismatch: expected nightly-cck-mint got nightly:	0.005%
+    Target channel mismatch: expected nightly-cck-rambler got nightly:	0.002%
+
+
+The vast majority of the data in the payload seems reasonable (99.71%). However, a handful of `update` pings are reporting a `targetBuildId` which is older than the current build reported by the ping's environment. Some pings are also reporting a `targetChannel` different than the one in the environment. These issues do not represent a problem, as most of the data is correct.
+
+## Check that we receive one ping per client and target update
+For each ping, build a key with the client id and the target update details. Since we expect to have exactly one ping for each update bundle marked as *ready*, we don't expect duplicate keys.
+
+
+```python
+update_dupes = deduped_subset.map(lambda p: ((p.get("clientId"),
+                                              p.get("payload/targetChannel"),
+                                              p.get("payload/targetVersion"),
+                                              p.get("payload/targetBuildId")), 1)).countByKey()
+
+print("Percentage of pings related to the same update (for the same client):\t{:.3f}%"\
+      .format(pct(sum([v for v in update_dupes.values() if v > 1]), deduped_count)))
+```
+    Percentage of pings related to the same update (for the same client):	1.742%
+
+
+Apparently we're receiving `update` pings with different `documentId` related to the same target update bundle, for a few clients. The 1.74% seems low enough to not be a problem at this stage.
```

Comment:
After cross checking this with Matt Howell, it seems that when updates fail to apply we could download them again. This would trigger a new update ping as well, with the same target update version/buildid. I've commented in the analysis.